File: mcp/src/main/java/top/emilejones/hhu/mcp/service/impl/RecallService.java
package top.emilejones.hhu.mcp.service.impl;

import kotlin.Pair;
import org.springframework.ai.tool.annotation.Tool;
import org.springframework.ai.tool.annotation.ToolParam;
import org.springframework.stereotype.Service;
import top.emilejones.hhu.mcp.entity.TextNode;
import top.emilejones.hhu.mcp.enums.TextType;
import top.emilejones.hhu.mcp.repository.IMilvusRepository;
import top.emilejones.hhu.mcp.repository.INeo4jRepository;
import top.emilejones.hhu.mcp.service.IRecallService;
import top.emilejones.hhu.model.ModelClient;
import top.emilejones.hhu.model.impl.XinferenceHttpClient;
import top.emilejones.hhu.model.pojo.RerankResult;
import top.emilejones.huu.env.XinferenceEnvironment;

import java.util.*;
import java.util.stream.Collectors;

/**
 * 用来负责处理召回任务
 *
 * @author EmileJones
 */
@Service
public class RecallService implements IRecallService {
    private IMilvusRepository milvusRepository;
    private INeo4jRepository neo4jRepository;
    private ModelClient client;

    public RecallService(IMilvusRepository milvusRepository, INeo4jRepository neo4jRepository) {
        this.milvusRepository = milvusRepository;
        this.neo4jRepository = neo4jRepository;
        client = new XinferenceHttpClient(XinferenceEnvironment.HOST, XinferenceEnvironment.PORT);
    }

    @Override
    @Tool(description = "根据问题返回和问题最相关的资料片段")
    public List<String> recallText(@ToolParam(description = "问题") String query) {
        final int maxResultNumber = 5;

        // 从向量数据库中召回数据
        List<Float> queryVector = client.embedding(query);
        List<TextNode> searchResults = milvusRepository.search(queryVector, 100);
        // 将召回的结果分批rerank（由于显存问题，需要分批rerank）
        final int step = 5;
        int index = 0;
        List<Pair<RerankResult, TextNode>> rerankResults = new ArrayList<>();

        while (index < searchResults.size()) {
            ArrayList<String> strings = new ArrayList<>();
            for (int i = 0; i < step; i++) {
                strings.add(searchResults.get(index + i).getText());
            }
            int startIndex = index;
            List<Pair<RerankResult, TextNode>> pairList = client.rerank(query, strings).stream().map(result -> {
                int index1 = result.getIndex();
                return new Pair<>(result, searchResults.get(index1 + startIndex));
            }).toList();
            rerankResults.addAll(pairList);
            index += step;
        }
        // 将分批rerank后的结果排序，获取得分最高的maxResultNumber个结果
        Set<TextNode> sets = rerankResults.stream()
                .sorted(Comparator.comparingDouble(value -> {
                    Pair<RerankResult, TextNode> value1 = (Pair<RerankResult, TextNode>) value;
                    return value1.getFirst().getScore();
                }).reversed())
                .limit(maxResultNumber)
                .map(obj -> {
                    Pair<RerankResult, TextNode> pair = (Pair<RerankResult, TextNode>) obj;
                    return pair.getSecond();
                })
                .collect(Collectors.toSet());
        // 将每一个节点向上向下查找，如果有表格上下文则加入
        Set<TextNode> resultSet = new HashSet<>();
        for (TextNode node : sets) {
            resultSet.add(node);
            // 向下找
            TextNode nowNode = node;
            while (nowNode.getType() == TextType.TABLE) {
                resultSet.add(nowNode);
                nowNode = neo4jRepository.nextNode(nowNode.getElementId());
                if (resultSet.contains(nowNode))
                    break;
            }
            // 向上找
            nowNode = node;
            while (nowNode.getType() == TextType.TABLE) {
                resultSet.add(nowNode);
                nowNode = neo4jRepository.preNode(nowNode.getElementId());
                if (resultSet.contains(nowNode))
                    break;
            }
        }

        return resultSet.stream().map(TextNode::getText).toList();
    }
}

File: infrastructure/textsplitter/structure-extraction/src/main/kotlin/top/emilejones/hhu/textsplitter/service/impl/RecallService.kt
package top.emilejones.hhu.textsplitter.service.impl

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import top.emilejones.hhu.domain.result.FileNode
import top.emilejones.hhu.infrastructure.configuration.env.pojo.RAGConfig
import top.emilejones.hhu.model.ModelClient
import top.emilejones.hhu.model.pojo.RerankResult
import top.emilejones.hhu.textsplitter.domain.po.neo4j.Neo4jTextNode
import top.emilejones.hhu.textsplitter.repository.IFileNodeMilvusRepository
import top.emilejones.hhu.textsplitter.repository.ITextNodeMilvusRepository
import top.emilejones.hhu.textsplitter.repository.INeo4jRepository
import top.emilejones.hhu.textsplitter.repository.impl.neo4j.extensions.asFileNode
import top.emilejones.hhu.textsplitter.service.IRecallService
import java.util.stream.Collectors

/**
 * @author EmileJones
 */
@Service
class RecallService(
    private val textNodeMilvusRepository: ITextNodeMilvusRepository,
    private val fileNodeMilvusRepository: IFileNodeMilvusRepository,
    private val neo4jRepository: INeo4jRepository,
    private val client: ModelClient,
    private val ragConfig: RAGConfig
) : IRecallService {

    override fun recallNode(query: String, collectionName: String, filter: String?): List<Neo4jTextNode> {
        val maxResultNumber = ragConfig.recallNumber

        // 从向量数据库中召回数据
        val queryVector: List<Float> = client.embedding(query)
        val searchResults = textNodeMilvusRepository.searchByVector(collectionName, queryVector, 100, filter)

        // 从Neo4j获取文本节点
        val neo4jNodes = neo4jRepository.batchSearchNeo4jTextNodeByNodeId(searchResults.map { it.neo4jNodeId })

        // 重排序结果，并取出得分最高的maxResultNumber个数据
        val rerankResult = client.rerank(query, neo4jNodes.map { it.text })
            .stream()
            .limit(maxResultNumber.toLong())
            .map { rr: RerankResult -> neo4jNodes[rr.index] }.toList()

        logger.debug(
            "查询问题[{}]召回节点的cypher语句为[{}]",
            query,
            generateCypherByNeo4jTextNodeList(rerankResult)
        )
        logger.info("用户问题为：[{}]，召回的节点数量为[{}]个", query, rerankResult.size)
        return rerankResult
    }

    override fun recallFileNode(query: String, collectionName: String): List<FileNode> = runCatching {
        val queryVector = client.embedding(query)

        // 1. 向量召回 30 个节点
        val searchResults = fileNodeMilvusRepository.searchByVector(
            collectionName = collectionName,
            queryVector = queryVector,
            topK = 30,
            filter = null
        )

        // 2. 从 Neo4j 获取节点详情
        val fileNodes = searchResults.mapNotNull { datum ->
            neo4jRepository.searchNeo4jFileNodeByNodeId(datum.fileNodeId)?.asFileNode()
        }

        if (fileNodes.isEmpty()) return emptyList()

        // 3. 准备重排序的文本列表
        val textList = fileNodes.map { it.fileAbstract ?: "" }

        // 4. 调用重排序模型进行重排序
        val rerankResults = client.rerank(query, textList)

        // 5. 根据重排序结果取 Top 5（ModelClient 已按分数降序排列）
        rerankResults
            .take(5)
            .map { fileNodes[it.index] }
    }.getOrDefault(emptyList())

    private fun generateCypherByNeo4jTextNodeList(Neo4jTextNodeList: List<Neo4jTextNode>): String {
        val list = Neo4jTextNodeList.stream()
            .map(Neo4jTextNode::id)
            .map { "\"" + it + "\"" }
            .collect(Collectors.joining(", ", "[", "]"))
        return "MATCH (n: TextNode) WHERE n.id IN %s MATCH (f:FileNode)-[:CONTAIN]->(n) RETURN n,f".format(
            list
        )
    }

    companion object {
        private val logger: Logger = LoggerFactory.getLogger(RecallService::class.java)
    }
}
